File: android/app/src/main/java/org/maturita/maturita/Route.kt
package org.maturita.maturita

sealed class Route(val page: String) {
    data object Welcome : Route("welcome")
    data object Subjects : Route("subjects")
    data object Stats : Route("stats")
    data object Roadmap : Route("roadmap")
    data class UnitMap(val unitId: Int) : Route("unit${unitId + 1}")
    data class GermanEx(val unitId: Int, val ex: Int) : Route("u${unitId + 1}e$ex")
    data class Vocab(val unitId: Int) : Route("u${unitId + 1}vocab")
    data object NetYears : Route("netyears")
    data object NetMap : Route("netmap")
    data class NetLesson(val id: Int) : Route("netunit$id")
    data class NetEx(val id: Int) : Route("netex$id")
    data object HwMap : Route("hwmap")
    data class HwLesson(val id: Int) : Route("hwunit$id")
    data class HwEx(val id: Int) : Route("hwex$id")
    data object CzechMap : Route("czechmap")
    data object Mluvnice : Route("mluvnice")
    data class MluvEx(val n: Int) : Route("mluve$n")
    data object ReadingList : Route("readinglist")
    data class Book(val id: String) : Route(if (id == "1984") "cetba1984" else "cetbaFuks")
    data class BookQuiz(val id: String) : Route(if (id == "1984") "cetba1984quiz" else "cetbaFuksQuiz")
    data class BookPlot(val id: String) : Route(if (id == "1984") "cetba1984dej" else "cetbaFuksDej")
}

fun routeFromPage(page: String): Route? = when (page) {
    "welcome" -> Route.Welcome
    "subjects" -> Route.Subjects
    "stats" -> Route.Stats
    "roadmap" -> Route.Roadmap
    "netyears" -> Route.NetYears
    "netmap" -> Route.NetMap
    "hwmap" -> Route.HwMap
    "czechmap" -> Route.CzechMap
    "mluvnice" -> Route.Mluvnice
    "readinglist" -> Route.ReadingList
    "cetba1984" -> Route.Book("1984")
    "cetba1984quiz" -> Route.BookQuiz("1984")
    "cetba1984dej" -> Route.BookPlot("1984")
    "cetbaFuks" -> Route.Book("fuks")
    "cetbaFuksQuiz" -> Route.BookQuiz("fuks")
    "cetbaFuksDej" -> Route.BookPlot("fuks")
    else -> {
        Regex("""unit(\d+)""").matchEntire(page)?.let {
            return Route.UnitMap(it.groupValues[1].toInt() - 1)
        }
        Regex("""u(\d+)e(\d+)""").matchEntire(page)?.let {
            return Route.GermanEx(it.groupValues[1].toInt() - 1, it.groupValues[2].toInt())
        }
        Regex("""u(\d+)vocab""").matchEntire(page)?.let {
            return Route.Vocab(it.groupValues[1].toInt() - 1)
        }
        Regex("""netunit(\d+)""").matchEntire(page)?.let {
            return Route.NetLesson(it.groupValues[1].toInt())
        }
        Regex("""netex(\d+)""").matchEntire(page)?.let {
            return Route.NetEx(it.groupValues[1].toInt())
        }
        Regex("""hwunit(\d+)""").matchEntire(page)?.let {
            return Route.HwLesson(it.groupValues[1].toInt())
        }
        Regex("""hwex(\d+)""").matchEntire(page)?.let {
            return Route.HwEx(it.groupValues[1].toInt())
        }
        Regex("""mluve(\d+)""").matchEntire(page)?.let {
            return Route.MluvEx(it.groupValues[1].toInt())
        }
        null
    }
}

File: ios/Maturita/Route.swift
import Foundation

enum AppTab: Hashable {
    case practice
    case progress
    case search
    case settings
}

enum Route: Hashable {
    case welcome
    case subjects
    case stats
    case roadmap
    case unitMap(Int)
    case germanEx(Int, Int)
    case vocab(Int)
    case netYears
    case netMap
    case netLesson(Int)
    case netEx(Int)
    case hwMap
    case hwLesson(Int)
    case hwEx(Int)
    case czechMap
    case mluvnice
    case mluvEx(Int)
    case readingList
    case book(String)
    case bookQuiz(String)
    case bookPlot(String)

    var page: String {
        switch self {
        case .welcome: return "welcome"
        case .subjects: return "subjects"
        case .stats: return "stats"
        case .roadmap: return "roadmap"
        case .unitMap(let unitId): return "unit\(unitId + 1)"
        case .germanEx(let unitId, let ex): return "u\(unitId + 1)e\(ex)"
        case .vocab(let unitId): return "u\(unitId + 1)vocab"
        case .netYears: return "netyears"
        case .netMap: return "netmap"
        case .netLesson(let id): return "netunit\(id)"
        case .netEx(let id): return "netex\(id)"
        case .hwMap: return "hwmap"
        case .hwLesson(let id): return "hwunit\(id)"
        case .hwEx(let id): return "hwex\(id)"
        case .czechMap: return "czechmap"
        case .mluvnice: return "mluvnice"
        case .mluvEx(let n): return "mluve\(n)"
        case .readingList: return "readinglist"
        case .book(let id):
            if id == "1984" { return "cetba1984" }
            if id == "fuks" { return "cetbaFuks" }
            return "cetba_" + id.replacingOccurrences(of: "-", with: "_")
        case .bookQuiz(let id):
            if id == "1984" { return "cetba1984quiz" }
            if id == "fuks" { return "cetbaFuksQuiz" }
            return "cetba_" + id.replacingOccurrences(of: "-", with: "_") + "_quiz"
        case .bookPlot(let id):
            if id == "1984" { return "cetba1984dej" }
            if id == "fuks" { return "cetbaFuksDej" }
            return "cetba_" + id.replacingOccurrences(of: "-", with: "_") + "_dej"
        }
    }

    /// Stack of destinations on top of the practice root.
    var stack: [Route] {
        switch self {
        case .welcome, .subjects, .stats:
            return []
        case .roadmap:
            return [.roadmap]
        case .unitMap(let id):
            return [.roadmap, .unitMap(id)]
        case .germanEx(let unit, let ex):
            return [.roadmap, .unitMap(unit), .germanEx(unit, ex)]
        case .vocab(let unit):
            return [.roadmap, .unitMap(unit), .vocab(unit)]
        case .netYears:
            return [.netYears]
        case .netMap:
            return [.netYears, .netMap]
        case .netLesson(let id):
            return [.netYears, .netMap, .netLesson(id)]
        case .netEx(let id):
            return [.netYears, .netMap, .netLesson(id), .netEx(id)]
        case .hwMap:
            return [.hwMap]
        case .hwLesson(let id):
            return [.hwMap, .hwLesson(id)]
        case .hwEx(let id):
            return [.hwMap, .hwLesson(id), .hwEx(id)]
        case .czechMap:
            return [.czechMap]
        case .mluvnice:
            return [.czechMap, .mluvnice]
        case .mluvEx(let n):
            return [.czechMap, .mluvnice, .mluvEx(n)]
        case .readingList:
            return [.czechMap, .readingList]
        case .book(let id):
            return [.czechMap, .readingList, .book(id)]
        case .bookQuiz(let id):
            return [.czechMap, .readingList, .book(id), .bookQuiz(id)]
        case .bookPlot(let id):
            return [.czechMap, .readingList, .book(id), .bookPlot(id)]
        }
    }
}

func routeFromPage(_ page: String) -> Route? {
    switch page {
    case "welcome": return .welcome
    case "subjects": return .subjects
    case "stats": return .stats
    case "roadmap": return .roadmap
    case "netyears": return .netYears
    case "netmap": return .netMap
    case "hwmap": return .hwMap
    case "czechmap": return .czechMap
    case "mluvnice": return .mluvnice
    case "readinglist": return .readingList
    case "cetba1984": return .book("1984")
    case "cetba1984quiz": return .bookQuiz("1984")
    case "cetba1984dej": return .bookPlot("1984")
    case "cetbaFuks": return .book("fuks")
    case "cetbaFuksQuiz": return .bookQuiz("fuks")
    case "cetbaFuksDej": return .bookPlot("fuks")
    default:
        if page.hasPrefix("cetba_"), page.hasSuffix("_quiz") {
            let slug = String(page.dropFirst(6).dropLast(5)).replacingOccurrences(of: "_", with: "-")
            return .bookQuiz(slug)
        }
        if page.hasPrefix("cetba_"), page.hasSuffix("_dej") {
            let slug = String(page.dropFirst(6).dropLast(4)).replacingOccurrences(of: "_", with: "-")
            return .bookPlot(slug)
        }
        if page.hasPrefix("cetba_") {
            let slug = String(page.dropFirst(6)).replacingOccurrences(of: "_", with: "-")
            return .book(slug)
        }
        if let m = page.wholeMatch(of: /unit(\d+)/), let n = Int(m.1) {
            return .unitMap(n - 1)
        }
        if let m = page.wholeMatch(of: /u(\d+)e(\d+)/), let u = Int(m.1), let e = Int(m.2) {
            return .germanEx(u - 1, e)
        }
        if let m = page.wholeMatch(of: /u(\d+)vocab/), let u = Int(m.1) {
            return .vocab(u - 1)
        }
        if let m = page.wholeMatch(of: /netunit(\d+)/), let n = Int(m.1) {
            return .netLesson(n)
        }
        if let m = page.wholeMatch(of: /netex(\d+)/), let n = Int(m.1) {
            return .netEx(n)
        }
        if let m = page.wholeMatch(of: /hwunit(\d+)/), let n = Int(m.1) {
            return .hwLesson(n)
        }
        if let m = page.wholeMatch(of: /hwex(\d+)/), let n = Int(m.1) {
            return .hwEx(n)
        }
        if let m = page.wholeMatch(of: /mluve(\d+)/), let n = Int(m.1) {
            return .mluvEx(n)
        }
        return nil
    }
}
